### ISSUE TITLE
Feature request: autobatching

### ISSUE THREAD

**Issue by beckdaniel** [neulab/xnmt#536]
Greetings,

I started using XNMT for my own research and from my understanding the current implementation does not benefit from autobatching, relying on explicit batching instead. From the autobatching tutorial, it seems that the only thing needed to enable it properly would be to sum losses for all instances in a batch before calling `forward()` in the computational graph, is that right? 

To give more context: I am currently implementing new encoders based on graph neural nets and these are very tricky to batch explicitly. Having autobatch would makes things a lot easier in this case. An alternative would be to accumulate calculations sequentially within my encoder, use explicit batching in the rest of the translator architecture and then trigger autobatch. However, I think that having autobatch in the whole architecture would make things more general and ease development of other components in the future.

How difficult would it be to implement this in XNMT? I am happy to give it a go if this is a reasonable request but any help is appreciated.

**Comment by neubig** [comment]
It depends what exactly you want to use autobatching for. Let's say you have some complicated encoder (e.g. https://arxiv.org/abs/1603.06075). If this is the case, you could just implement this encoder by doing a for loop over the (e.g. tree-structured) inputs, and then combine everything together into one batched xnmt `ExpressionSequence`. Then if you turned on the autobatching flag, the first part using the complicated encoder would benefit from autobatching, while the decoder would benefit from explicit batching.

I think @philip30 's segmenting encoder is based on a concept like this, so it could provide an example: https://github.com/neulab/xnmt/blob/master/xnmt/specialized_encoders/segmenting_encoder/segmenting_encoder.py

Would this suit your use case?

**Comment by msperber** [comment]
The approach at the loss level would also be possible even right now with a hack by using the ```SameBatchMultiTaskTrainingRegimen``` with ```per_task_backward=False``` and then simply specifying the same task multiple times (as many as there should be elements in the minibatch):
https://xnmt.readthedocs.io/en/latest/api_doc.html#xnmt.train.regimens.SameBatchMultiTaskTrainingRegimen

That's of course a bit awkward, but I think it should be simple to write a new training regimen (or extend ```SimpleTrainingRegimen```) with a similar for loop that simply goes over the same task several times.

**Comment by beckdaniel** [comment]
Hi @msperber,
If my understanding of the code is correct, `SameBatchMultiTaskTrainingRegimen with `per_task_backward=False` would only delay calls to `backward()`. The forward pass is still done sequentially when calling `loss_builder.compute()`, no? Doesn't autobatching require the forward pass to be accumulated as well?

**Comment by msperber** [comment]
```loss_builder.compute()``` doesn't actually trigger a forward pass, it only aggregates DyNet Expressions. So as far as I can see this should still work because the forward pass is also delayed accordingly (```per_task_backward``` may not have been the best naming choice).

**Comment by beckdaniel** [comment]
I see. But then, which part of the code triggers the forward pass? For instance, in the method `run_training` in `SimpleTrainingRegimen`, `loss_builder.compute()` is right before `self.backward(loss)`. I might be mistaken but the forward pass should be done before calling `backward()`, no?

Meanwhile I decided to give it a go by having a new  `AutobatchTrainingRegimen`, which uses the `update_every` keyword to manage losses:
https://github.com/beckdaniel/xnmt/blob/autobatch/xnmt/train/regimens.py
It is still buggy (crashes when batch_size is not a multiple of the dataset size) and I am not sure if it is doing the right thing.

I might also try the multitask approach but I need to understand how tasks work in XNMT first.

**Comment by msperber** [comment]
In my understanding DyNet will call the forward pass lazily, i.e. whenever backward is invoked on a loss expression and the corresponding forward values don't exist yet. So that means ```self.backward(loss)``` will implicitly call the forward pass as well. The forward pass can also be performed explicitly by calling ```some_expression.value()```. ```value()``` is actually used in XNMT for purposes of logging the loss values (```self.train_loss_tracker.report()```). However, this happens only after the complete update is performed, at which point ```value()``` will simply return the stored value that had been computed previously.

**Comment by msperber** [comment]
Also, just briefly looking at your code I think it's doing the right thing as far as the calls to ```backward()``` and ```update()``` are concerned.

**Comment by beckdaniel** [comment]
Ah you're right, it's just that most DyNet tutorials call `npvalue()` explicitly.

However, in my code, I think the loss tracker still calls forward explicitly, by calling `get_factored_loss_val` in `losses.py`, no? At least I am not seeing too many speed gains from autobatching at the moment.

I'll see if I can come up with a proper example of this, using one of the YAML templates in the examples folder.

**Comment by msperber** [comment]
I see, yeah that would likely be the cause. So the loss tracker should definitely only be called after backward in this case. And you could accumulate ```FactoredLossExpr``` loss expressions within the minibatch by using ```add_factored_loss_expr()```, instead of keeping a list of losses. That way you can pass the accumulated loss expression to the loss tracker after the update.

**Comment by beckdaniel** [comment]
Hi, sorry for being off the radar for a while.

I added two YAML examples, trying to benchmark my autobatching implementation. These are similar to `01_standard.yaml`, except that they use the "big" files already present in the `examples/data` folder. The corresponding YAML files are `23a_big_model.yaml` and `23b_big_model_autobatch.yaml`, here:

https://github.com/beckdaniel/xnmt/tree/autobatch/examples

At the moment, I'm getting these speeds when running on a Titan X GPU:

- `batch_size` 32, standard batching and training regimen:
`[big] Epoch 0.0069: train_loss/word=7.844767 (words=9434, words/sec=3009.57, time=0-00:00:12)`
`[big] Epoch 0.0138: train_loss/word=7.266208 (words=19624, words/sec=3025.77, time=0-00:00:16)`
`[big] Epoch 0.0206: train_loss/word=6.946810 (words=29420, words/sec=3097.93, time=0-00:00:19)`
`[big] Epoch 0.0275: train_loss/word=6.746871 (words=38777, words/sec=2979.56, time=0-00:00:22)`
`[big] Epoch 0.0344: train_loss/word=6.619986 (words=49056, words/sec=3027.75, time=0-00:00:26)`
`[big] Epoch 0.0413: train_loss/word=6.507618 (words=58298, words/sec=3011.75, time=0-00:00:29)`
`[big] Epoch 0.0482: train_loss/word=6.403921 (words=67647, words/sec=3023.31, time=0-00:00:32)`

- `batch_size` 1, `update_every` 32, autobatching *on* + new autobatching training regimen:
`[big_autobatch] Epoch 0.0069: train_loss/word=7.829281 (words=9579, words/sec=785.80, time=0-00:00:22)`
`[big_autobatch] Epoch 0.0138: train_loss/word=7.274475 (words=19293, words/sec=769.66, time=0-00:00:35)`
`[big_autobatch] Epoch 0.0206: train_loss/word=7.002695 (words=28748, words/sec=772.72, time=0-00:00:47)`
`[big_autobatch] Epoch 0.0275: train_loss/word=6.799202 (words=38201, words/sec=787.97, time=0-00:01:00)`
`[big_autobatch] Epoch 0.0344: train_loss/word=6.636837 (words=47833, words/sec=756.96, time=0-00:01:13)`
`[big_autobatch] Epoch 0.0413: train_loss/word=6.516927 (words=57457, words/sec=772.95, time=0-00:01:25)`
`[big_autobatch] Epoch 0.0482: train_loss/word=6.409992 (words=67078, words/sec=780.42, time=0-00:01:38)`

- `batch_size` 1, `update_every` 32, autobatching *off* + new autobatching training regimen:
`[big_autobatch] Epoch 0.0069: train_loss/word=7.808652 (words=9610, words/sec=261.05, time=0-00:00:47)`
`[big_autobatch] Epoch 0.0138: train_loss/word=7.272679 (words=19048, words/sec=258.80, time=0-00:01:24)`
`[big_autobatch] Epoch 0.0206: train_loss/word=6.969558 (words=28773, words/sec=259.86, time=0-00:02:01)`
`[big_autobatch] Epoch 0.0275: train_loss/word=6.763105 (words=38323, words/sec=259.87, time=0-00:02:38)`
`[big_autobatch] Epoch 0.0344: train_loss/word=6.612972 (words=47791, words/sec=258.67, time=0-00:03:15)`
`[big_autobatch] Epoch 0.0413: train_loss/word=6.482694 (words=57277, words/sec=258.50, time=0-00:03:52)`
`[big_autobatch] Epoch 0.0482: train_loss/word=6.379741 (words=66808, words/sec=260.39, time=0-00:04:29)`

Hopefully these logs are not too confusing, let me know you need clarification. But the main point is: 1) t using autobatching gives around 3x speedup at the moment, but still is 4x slower compared to standard batching. Is my implementation missing something or am I reaching the limits of what you can do autobatching?

**Comment by neubig** [comment]
These results seem pretty reasonable to me. There are two reasons why autobatching may be slower than manual batching:
1. Python overhead for running the graph expansion N times instead of just once (where N is the size of the batch)
2. The autobatching algorithm itself not finding all of the batching opportunities.

In this case I suspect the great majority is "1.", so running a Python profiler and finding the places that are taking a lot of time might be one way to bring autobatching closer to manual batching (and speed things up in general).

**Comment by beckdaniel** [comment]
Thanks for the feedback. I'll see what I can get from a profiler then.

Should I start a pull request on this?

**Comment by beckdaniel** [comment]
Should I close this, at least for now?

**Comment by philip30** [comment]
Yea, we should close the issue after the PR is merged. Don't forget to refer the issue in the PR :)
